File: daily_challenges/3542.cpp
//T.C : O(n*u), u = unique elements, in worst case  u = n
//S.C : O(u), in worst case u = n
// Gave TLE
class Solution {
public:
    int minOperations(vector<int>& nums) {
        unordered_set<int> st(begin(nums), end(nums)); //O(n) space

        int n = nums.size();

        int ops = 0;

        for(int target : st) { //O(U*n)
            if(target == 0)
                continue;
            
            bool flow = false;
            for(int i = 0; i < n; i++) {
                if(nums[i] == target) {
                    if(!flow) {
                        flow = true;
                        ops++;
                    }
                } else if(nums[i] < target) {
                    flow = false;
                }
            }
        }

        return ops;

        
    }
};


// Monotonic Increasing Stack
//T.C : O(n)
//S.C : O(n)

File: daily_challenge_leetcode/3542.py


#  Monotonic Increasing Stack
#  T.C : O(n)
#  S.C : O(n)

from typing import List
class Solution:
    def minOperations(self, nums: List[int]) -> int:
        st = []
        ops = 0
        for num in nums:
            while st and st[-1] > num:
                st.pop()
            if num == 0:
                continue
            if not st or st[-1] < num:
                st.append(num)
                ops += 1
        return ops
